src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,328 @@
+
+            if (hbeat.Properties.ContainsKey(HeartbeatDefaultPayload.UpdatedFieldsPropertyKey))
+            {
+                updatedKeys = hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey];
+                comma = ",";
+            }
+
+            foreach (var payloadItem in this.heartbeatProperties)
+            {
+                hbeat.Properties.Add(payloadItem.Key, payloadItem.Value.PayloadValue);
+                hbeat.Sum += payloadItem.Value.IsHealthy ? 0 : 1;
+                if (payloadItem.Value.IsUpdated)
+                {
+                    string.Concat(updatedKeys, comma, payloadItem.Key);
+                    comma = ",";
+                }
+            }
+
+            // update the special 'updated keys' property with the names of keys that have been updated.
+            if (!string.IsNullOrEmpty(updatedKeys))
+            {
+                hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey] = updatedKeys;
+            }
+
+            hbeat.Sequence = string.Format(CultureInfo.CurrentCulture, "{0}", this.heartbeatsSent++);
+
+            return hbeat;
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                    if (this.HeartbeatTimer != null)
+                    {
+                        this.isEnabled = false;
+
+                        try
+                        {
+                            this.HeartbeatTimer.Dispose();
+                        }
+                        catch (Exception)
+                        {
+                            CoreEventSource.Log.LogError("Error occured when disposing heartbeat timer within HeartbeatProvider");
+                        }
+                    }
+                }
+
+                this.disposedValue = true;
+            }
+        }
+
+        #endregion
+
+        private void Send()
+        {
+            if (this.telemetryClient.TelemetryConfiguration.TelemetryChannel == null)
+            {
+                return;
+            }
+
+            var eventData = (MetricTelemetry)this.GatherData();
+
+            eventData.Context.Operation.SyntheticSource = heartbeatSyntheticMetricName;
+
+            eventData.Context.InstrumentationKey = this.InstrumentationKey;
+
+            this.telemetryClient.TrackMetric(eventData);
+        }
+
+        private void HeartbeatPulse(object state)
+        {
+            if (state is HeartbeatProvider)
+            {
+                HeartbeatProvider hp = state as HeartbeatProvider;
+                // we will be prone to overlap if any extension payload provider takes a longer time to process than our timer
+                // interval. Best that we reset the timer each time round.
+                try
+                {
+                    this.HeartbeatTimer.Change(Timeout.Infinite, Timeout.Infinite);
+
+                    if (this.isEnabled)
```

Comment:
so first tick off timer in disabled state will turn off the feature permanently?

Reply:
Timer will not fire as you'd not change the interval back

Reply:
No, see the property ‘IsEnabled’ for the on/off logic here.

Setting IsEnabled=false will disable the timer. Setting IsEnabled=true will enable the timer. This logic here is where a tick is ongoing (on a different thread) while a user has set IsEnabled=false, unlikely scenario but if we don’t plan for it we will certainly get bitten, once we make this generally available to users via some API.